components/textbox/textbox.tsx

```diff
@@ -66,257 +67,277 @@ type Props = {
-                priorityProfiles: this.props.priorityProfiles,
+                currentUserId,
+                channelId,
+                autocompleteUsersInChannel: (prefix: string) => actions.autocompleteUsersInChannel(prefix, channelId),
+                useChannelMentions,
+                autocompleteGroups,
+                searchAssociatedGroupsForReference: (prefix: string) => actions.searchAssociatedGroupsForReference(prefix, currentTeamId, channelId),
+                priorityProfiles,
             }),
-            new ChannelMentionProvider(props.actions.autocompleteChannels),
+            new ChannelMentionProvider(actions.autocompleteChannels),
             new EmoticonProvider(),
         );
 
-        if (props.supportsCommands) {
-            this.suggestionProviders.push(new CommandProvider({
-                teamId: this.props.currentTeamId,
-                channelId: this.props.channelId,
-                rootId: this.props.rootId,
-            }));
-        }
-
-        this.checkMessageLength(props.value);
-        this.wrapper = React.createRef();
-        this.message = React.createRef();
-        this.preview = React.createRef();
-    }
-
-    handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
-        this.props.onChange(e);
-    }
+        updateSuggestionProviders(newProviders);
+    }, [autocompleteGroups, channelId, currentTeamId, currentUserId, actions, priorityProfiles, useChannelMentions, rootId, supportsCommands]);
 
-    updateSuggestions(prevProps: Props) {
-        if (this.props.channelId !== prevProps.channelId ||
-            this.props.currentUserId !== prevProps.currentUserId ||
-            this.props.autocompleteGroups !== prevProps.autocompleteGroups ||
-            this.props.useChannelMentions !== prevProps.useChannelMentions ||
-            this.props.currentTeamId !== prevProps.currentTeamId ||
-            this.props.priorityProfiles !== prevProps.priorityProfiles) {
+    const checkMessageLength = useCallback((message: string) => {
+        if (handlePostError) {
+            if (message.length > characterLimit) {
+                const errorMessage = (
+                    <FormattedMessage
+                        id='create_post.error_message'
+                        defaultMessage='Your message is too long. Character count: {length}/{limit}'
+                        values={{
+                            length: message.length,
+                            limit: characterLimit,
+                        }}
+                    />);
+                handlePostError(errorMessage);
+            } else {
+                handlePostError(null);
+            }
+        }
+    }, [characterLimit, handlePostError]);
+
+    const updateSuggestions = useCallback((prevProps: Pick<Props, 'channelId' | 'currentUserId' | 'autocompleteGroups' | 'useChannelMentions' | 'currentTeamId' | 'priorityProfiles' | 'rootId' | 'value'>) => {
```

Comment:
I'd like to move this to outside the component as well, because it's more business logic. I had some issues with rendering by doing that but I'm sure it can be done.